kyo-tag/shared/src/main/scala/kyo/tags.scala

```diff
@@ -2,110 +2,129 @@ package kyo
-        end extension
+        infix def =:=[U](t2: Full[U]): Boolean =
+            t2 match
+                case t2: Union[?] =>
+                    tags.forall(tag1 => t2.tags.exists(tag1 =:= _)) &&
+                    t2.tags.forall(tag2 => tags.exists(tag2 =:= _))
+                case _ =>
+                    tags.forall(t2 =:= _)
     end Union
 
-    case class Intersection[T](tags: Seq[Tag[?]]) extends AnyVal
-
-    object Intersection:
-
-        inline given apply[T]: Intersection[T] = ${ intersectionImpl[T] }
-
-        extension [T](t1: Intersection[T])
-            def show: String =
-                t1.tags.map(_.show).mkString(" & ")
-
-            infix def <:<[U](t2: Tag[U]): Boolean =
-                t1.tags.exists(_ <:< t2)
-
-            infix def =:=[U](t2: Tag[U]): Boolean =
-                t1.tags.forall(_ =:= t2)
-
-            infix def =!=[U](t2: Tag[U]): Boolean =
-                !(t1 =:= t2)
+    object Union:
+        private[Tag] def raw[T](tags: Seq[String]) = new Union[T](tags.asInstanceOf[Seq[Tag[Any]]])
+        inline given apply[T]: Union[T]            = ${ unionImpl[T] }
 
-            infix def >:>[U](t2: Tag[U]): Boolean =
-                t1.tags.forall(t2 >:> _)
+    case class Intersection[T](tags: Seq[Tag[Any]]) extends Set[T]:
+        def show: String =
+            tags.map(_.show).mkString(" & ")
 
-            infix def <:<(t2: Intersection[?]): Boolean =
-                t2.tags.forall(tag2 => t1.tags.exists(_ <:< tag2))
+        infix def <:<[U](t2: Full[U]): Boolean =
+            t2 match
+                case t2: Intersection[?] =>
+                    t2.tags.forall(tag2 => tags.exists(_ <:< tag2))
+                case _ =>
+                    tags.exists(_ <:< t2)
 
-            infix def =:=(t2: Intersection[?]): Boolean =
-                t1.tags.forall(tag1 => t2.tags.exists(tag1 =:= _)) &&
-                    t2.tags.forall(tag2 => t1.tags.exists(tag2 =:= _))
+        infix def >:>[U](t2: Full[U]): Boolean =
+            t2 match
+                case t2: Intersection[?] =>
+                    tags.forall(tag1 => t2.tags.exists(tag1 >:> _))
+                case _ =>
+                    tags.forall(t2 >:> _)
 
-            infix def =!=(t2: Intersection[?]): Boolean =
-                !(t1 =:= t2)
+        infix def =:=[U](t2: Full[U]): Boolean =
+            t2 match
+                case t2: Intersection[?] =>
+                    tags.forall(tag1 => t2.tags.exists(tag1 =:= _)) &&
+                    t2.tags.forall(tag2 => tags.exists(tag2 =:= _))
+                case _ =>
+                    tags.forall(_ =:= t2)
 
-            infix def >:>(t2: Intersection[?]): Boolean =
-                t1.tags.forall(tag1 => t2.tags.exists(tag1 >:> _))
-        end extension
     end Intersection
 
+    object Intersection:
+        private[Tag] def raw[T](tags: Seq[String]) = new Intersection[T](tags.asInstanceOf[Seq[Tag[Any]]])
```

Comment:
I was having issues with the macro generating invalid trees when I changed `Tag` to be `opaque`. This is a workaround for it.